File: 2024/day2.py
import AOCInit
import util

if __name__ != '__main__':
    exit()

inp = util.getInput(d=2, y=2024)

nums = tuple(util.getInts(line) for line in inp.splitlines())

# part 1
def isSafeRow(row):
    n = len(row)
    incFail = False
    decFail = False
    for i in range(1, n):
        if row[i] - row[i - 1] not in range(1, 4):
            incFail = True
        if row[i - 1] - row[i] not in range(1, 4):
            decFail = True
        if incFail and decFail:
            return False
    return True

states = []
for row in nums:
    states.append(isSafeRow(row))
print(sum(states))


# part 2
count = 0
for i, row in enumerate(nums):
    if not states[i]:
        n = len(row)
        for j in range(n):
            if isSafeRow(row[:j] + row[j + 1:]):
                states[i] = True
                break
print(sum(states))

File: 2024/day2.cpp
#include <iostream>
#include <string>
#include <vector>

#include "../util.h"

using ulli = unsigned long long int;

bool isSafeRow(const std::vector<int> &row){
    int n = row.size();
    bool incFail = false, decFail = false;
    for (int i = 1; i < n; ++i){
        int delta = row[i] - row[i - 1];
        if (!(1 <= delta && delta <= 3)){
            incFail = true;
        }
        if (!(-3 <= delta && delta <= -1)){
            decFail = true;
        }
        if (incFail && decFail){
            return false;
        }
    }
    return true;
}

int main(int, char**){
    std::string inp = util::getInput(2024, 2);
    std::vector<std::vector<int>> nums;
    for (auto &s : util::splitline(inp)){
        nums.push_back(util::getInts(s));
    }

    // part 1
    int n = nums.size();
    std::vector<bool> passState(n, false);
    int res = 0;
    for (int i = 0; i < n; ++i){
        res += passState[i] = isSafeRow(nums[i]);
    }
    std::cout << res << std::endl;


    // part 2
    res = 0;
    for (int i = 0; i < n; ++i){
        if (!passState[i]){
            std::vector<int> buff(nums[i].size() - 1);
            for (int j = 0; j < nums[i].size(); ++j){
                int wptr = 0;
                for (int k = 0; k < nums[i].size(); ++k){
                    if (k == j){
                        continue;
                    }
                    buff[wptr++] = nums[i][k];
                }
                if (isSafeRow(buff)){
                    passState[i] = true;
                    break;
                }
            }
        }
        res += passState[i];
    }
    std::cout << res << std::endl;

    return 0;
}
